reference/plugins.md

```diff
@@ -2,7 +2,170 @@
+        unless Plugin.authenticate(controller.request)
+          return Praxis::Responses::Unauthorized.new(body: 'unauthorized')
+        end
+      end
+    end
+  end
+
+end
+{% endhighlight %}
+
+We add a handy `requires_authentication` method to action definitions in the `ActionDefinition` module.
+{% highlight ruby %}
+module ActionDefinition
+  # and again, extend AS::Concern. This time so we can use
+  # decorate_docs in the proper context.
+  extend ActiveSupport::Concern
+
+
+  included do
+    # add an :authentication_required key to the generated documentation.
+    # note: this is only for the generated JSON output, handling the display
+    # in the doc browser is separate.
+    decorate_docs do |action, docs|
+      docs[:authentication_required] = action.authentication_required
+    end
+  end
+
+  # simple requires_authentication dsl helper.
+  def requires_authentication(value)
+    @authentication_required = value
+  end
+
+  # and a quick authentication_required getter, defaulting to false.
+  def authentication_required
+    @authentication_required ||= false
+  end
+
+end
+{% endhighlight %}
+
+
+
+## Using a Plugin
+
+In order to use a plugin named `MyPlugin`, you would typically have the following in your `config/environment.rb`:
+
+{% highlight ruby %}
+Praxis::Application.configure do |application|
+  application.bootloader.use MyPlugin, **my_plugin_options
+end
+{% endhighlight %}
+
+
+
+As outlined above, `MyPlugin` may be either a subclass of `Plugin` *or* a module that includes `PluginConcern`. In the latter case, Praxis will expect there to be a class named `Plugin`, e.g. `MyPlugin::Plugin`.
+
+
+## Plugin Components
+
+The two main components of a plugin in Praxis are:
+
+* `Plugin`: The class that is instantiated for each use of the plugin (unless it includes `Singleton`, in which case its `instance` is used).
+  * `Plugin#config_key`: the name used by Praxis to store the plugin's instance the `Application#plugins` hash, and where relevant configuration data will be located in `Application#config`.
+* `PluginConcern`: An `ActiveSupport::Concern` module that encloses extensions to core Praxis classes.
+
+A plugin in Praxis *must* consist of a subclass of `Plugin`, and that subclass *may* be enclosed in a module that has included `PluginConcern`
+
+## Plugin Bootstrapping and Configuration
+
+When `Bootloader#use` is called when loading an application's `environment.rb`, a number of things happen:
+
+1. If the plugin provided is a module that includes `PluginConcern`, that module's `setup!` method is called. By default, that method will inject any relevant modules into the core Praxis classes, provided that is the first time the method has been called (in the event the plugin is used multiple times in one application).
+2. An instance of that plugin's `Plugin` subclass is created and saved in the application's set of plugins. Or, if that subclass is a `Singleton`, then its `instance` is retrieved and used instead.
```

Comment:
Are you meaning subclass in the real sense? I think maybe you just mean the Plugin class within the plugin's namespace, right? and this only happens if there is a Plugin class I think. That might be worth noting here.